### ISSUE TITLE
[Unintended Issue?] Values below 0.01 are reset to 0.01

### ISSUE THREAD

**Issue by VasVadum** [thommcgrath/Beacon#42]
`"EntryWeight":0.001,"Items":[{"Class":"PrimalItemResource_GloryCoin_C","Weight":1.0}],`
became
`"EntryWeight":0.010000000000000000208,"Items":[{"Class":"PrimalItemResource_GloryCoin_C","Weight":1.0}],`

after I re-saved the file.  I had manually edited the .beacon file to get my value that low in the first place.

Also there are strangeish issues where the slider shows a different value.  Like if I slide over to "69%", I expect 0.69, right?  But instead I get "0.68999999999999994671".  Every time.  I suppose it comes out to be the same thing anyway, but just another strange thing I noticed.  :P

**Comment by thommcgrath** [comment]
For the reset, I'll have to check into that. For the strange number, it's because computer's don't really like decimals. A floating point number is really a fraction. A numerator and denominator. .69 can't be represented nicely as a fraction, so the system has to choose something that CAN be represented as a fraction. It's not my doing, just something the compiler does and nothing I can really control, aside from rounding to a lower number of decimals.

**Comment by VasVadum** [comment]
I'm having great difficulty trying to set weights below 1% xD  I had to export, so I could edit the ini file directly.  Now everythings set to 0.01, and trying to figure out how to convert that to 0.25% and such.  :P Definately hope to see decimal values allowed at some point.

**Comment by thommcgrath** [comment]
I really think you're overthinking this. Stop thinking of it in terms of percentages. Even if you have 5 items with a weight of 0.01, each will still have a chance of 20%. Because 0.01 / 0.05 = 0.2. The weights are relative to each other.

**Comment by VasVadum** [comment]
Then that requires me to do all new extremely difficult math if I don't think of each one as a percentage. I can't achieve what I'm trying to get if I need to do some physics and orbital dynamics and astrophysics equations first!  The method I'm doing now looks like it kinda works the way I want it to.  Its not perfect but eh, close enough.

**Comment by thommcgrath** [comment]
Well, no, it requires you do a bunch of math in order to think of it as a percentage. If you think about it as weight, these problems basically go away. An item with a weight of 100 is 100 times more likely to be selected as an item with 1 weight. An item with 50 weight is 2 times more likely to be selected as an item with 25 weight.

I could allow <1 weights, but it won't achieve what you're looking for. It just isn't a percentage. I cannot allow a specific chance to be input for two reasons. Each part is exported by itself and joined at the end. Set Entry A has no idea Set Entry B exists, so the information Set Entry A needs to convert a percentage to a weight isn't there since it would need to know of all its siblings. Plus, what if it is configured wrong? Two items with 20% chance cannot compute correct values.

**Comment by VasVadum** [comment]
Well, are you able to stop the program from resetting values below 0.01? I'm getting coins way too often in my beacons that were all set to 1%, almost every 5th beacon or so.  I'll show you my beacon file here soon if you want.  I'm actually trying to set up an automation thing, that works with your beacon tool.  Any chance you can help me figure out how to go about this?

https://docs.google.com/spreadsheets/d/1xodurE_mQyDvq2dx6hQqoKHHc16yycuAIttdhHuRIvA/pubhtml?gid=191463038&single=true

You can see that the last cell in row 5, is bold, and contains part of the string for controlling the item set.  In it, is [] but I'm not sure how to go about having it add all the strings below it into that without using a whole bunch more concat(X,X) inside it.  I'm doing this because its much more appealing and easier to edit, for me, than using the beacon editor which keeps rearranging things I add to it.  If I can find a way to get this document to output what I need, then I can build a whole beacon file from right here, then have your tool export the beacon file as overrides.

Theres another question I have too, since I'm gonna be making a mod soon that sorta goes with your beacon editor a little bit.  I'll ask it later though, gonna take more than a week I think to get Ark Dev Kit.

**Comment by thommcgrath** [comment]
Ok, regarding your document, add a row with the JOIN function. First parameter should be ",", second is the range. Surround that with [ and ] and you should have a JSON string.

I have not made this specific change yet, as I've been working on a truly massive update. But I'll make you a special build with this change in place, you can test it out, and tell me if it makes a difference. I suspect it won't though, something tells me this is Ark being weird. If you're seeing coins that frequently, even at the weights Beacon currently outputs, I think something else is going on.

You're on Windows right? Just so I know which platform to build for.

**Comment by VasVadum** [comment]
Yea, I'm on Windows.  I am patient, I can wait for the major update if you prefer.  Rather it be easier for you.  I've actually got a few ideas now as well.

You know how you can copy one entry from one beacon window to another beacon window?  What's the past data? I tried to paste into Notepad, to see what it looked like.  Was hoping it was something I might be able to concat/join in the document, then paste right into the beacon tool.

I am getting some other odd issues as well with in game beacons, some seem to give duplicate things when I don't want them to even after being set not to.  I could take this to another issue instead if you prefer.  If you got something like discord/steam/telegram/whatever we can discuss these things when you're around and such.  I am about to re-work my beacons taking into account weight not being percentage. Just trying to automate my spreadsheet, so I can keep my current nice looking setup that lets me see the data in a manner I need.

**Comment by VasVadum** [comment]
Here's an example of some things from your file I don't understand;

      {
      "Availability": 1, -- What?
      "bSetsRandomWithoutReplacement": true,
      "ItemSets": [
        {{AUTOMATED}} -- The two sections I cut out already
      ],
      "Kind": "Standard", -- What?
      "Label": "Island White (Level 3)",
      "MaxItemSets": 1,
      "MinItemSets": 1,
      "Multiplier_Max": 1.0, -- What?
      "Multiplier_Min": 1.0, -- What?
      "NumItemSetsPower": 1.0, -- What?
      "SortValue": 0, -- What?
      "SupplyCrateClassString": "SupplyCrate_Level03_C",
      "UIColor": "FFFFFF00"
    },

I suppose because those things don't appear to be in use at all, I guess there's no point in me creating separate fields for them? :P Unless your future major update, plans to use them?

**Comment by thommcgrath** [comment]
No, I'd rather get your help with this "issue" before release. It's a piece of cake for me to go back to the old code and make changes. So I've prepared beta "7.1" which has literally one change: the allowed range for an item set weight is 0-1 instead of 0.01-1. That should allow you freedom to experiment some more. Problem is GitHub won't let me attach it here, as it's too big. So I need to figure out a better way.

I've setup a Discord server I'll start offering as a support option in the future, we'll likely need to share the file that way https://discord.gg/2vbT7fV

When you copy a loot source, there is private data for Beacon and a text version ready to be pasted into an ini. Everything else uses private data that cannot be pasted elsewhere.

**Comment by thommcgrath** [comment]
- "Kind" tells Beacon which icon to display.
- The multipliers are the hidden quality multipliers built into the game. The only reason they are stored in the file is in case the loot source is a custom one provided by a mod map, the values will not be lost when saved.
- NumItemSetsPower is a value that could probably be removed, it's used by Ark and Wildcard has told us to keep it at 0. Its existence is there to maintain imported configs.
- Sort is used by Beacon to instruct it how to organize sources.
- UIColor tells Beacon how to color the icon.

I believe I wrote Beacon to gracefully degrade if any of those are missing, but I'm not 100% certain right now. I don't think I anticipated users modifying the file directly.
